Improving documentation regarding full row format and operation codes (#1565)
redis/docs: content/integrate/redis-data-integration/data-pipelines/transform-examples/redis-row-format.md

---
Title: Row Format

> aliases: /integrate/redis-data-integration/ingest/data-pipelines/transform-examples/redis-row-format/

Reviewer: The aliases are only required for files that have moved location (ie, it points to the old location of the file to help with redirection).
```suggestion
```

alwaysopen: false
categories:
- docs
- integrate
- rs
- rdi
description: null
group: di
linkTitle: Row Format
summary: Explanation of the row formats supported by Redis Data Integration jobs.
type: integration
weight: 30
---

> The RDI pipelines support two separate row formats which can be specified in the `source` section of the job file:
> 
> - `basic` - this is the default one, when no `row_format` is specified. It contains only the current value of the row
> - `full` - you get a full representation of the row, including the key, the before and after values, the operation code, etc.

Reviewer: ```suggestion
The RDI pipelines support two separate row formats which you can specify in the `source` section of the job file:

- `basic` - (Default) Contains the current value of the row only.
- `full` - Contains all information available for the row, including the key, the before and after values, and the operation code.
```

> The `full` row format is useful when you want to access the metadata associated with the row, such as the operation code, the before and after values, etc.
> Based on the used row format, the structure of the data passed to the `transform` and `output` sections is different and the keys should be referenced accordingly to the chosen row format.

Reviewer: ```suggestion
The `full` row format is useful when you want to access the metadata associated with the row, such as the operation code, and the before and after values.
The structure of the data passed to the `transform` and `output` sections is different depending on the row format you choose. Consider which row format you are using when you reference keys.
```

The following two examples demonstrate the difference between the two row formats.

## Default row format

With the default row format, the input value is a JSON object containing the current value of the row, and fields can be referenced directly by their name.

Usage example:

```yaml
source:
  table: addresses
transform:
  - uses: add_field
    with:
      field: city_state
      expression: concat([CITY, ', ', STATE])
      language: jmespath
  - uses: add_field
    with:
      field: op_code_value
      # Operation code is not available in standard row format
      # so the following expression will result in `op_code - None`
      expression: concat(['op_code', ' - ', opcode])
      language: jmespath
output:
  - uses: redis.write
    with:
      data_type: hash
      key:
        expression: concat(['addresses', '#', ID])
        language: jmespath
```

## Full row format {#full}

With `row_format: full` the input value is a JSON object with the following structure:

> - `key` - An object structure containing the attributes of the primary key. For example, `key.id` will give you the value of it the `id` column as long as it is part of the primary key.
> - `before` - An object structure containing the previous value of the row.
> - `after` - An object structure containing the current value of the row.
> - `opcode` - The operation code. Please note that some databases use different values for the operation code. Please refer to the [operation code values]({{< relref "#operation-codes" >}}) section for more information.

Reviewer: ```suggestion
- `key` - An object containing the attributes of the primary key. For example, `key.id` will give you the value of the `id` column as long as it is part of the primary key.
- `before` - An object containing the previous value of the row.
- `after` - An object containing the current value of the row.
- `opcode` - The operation code. Different databases use different values for the operation code. See [operation code values]({{< relref "#operation-codes" >}}) below for more information.
```
